models/camelot/camelot_pair_fee_rates.sql

```diff
@@ -0,0 +1,163 @@
+    ),
+    v3_pairs_with_initial_fee_rates as (
+        select
+            date_trunc('minute', evt_block_time) as minute,
+            pool as pair,
+            '3' as version,
+            token0,
+            {{ v3_default_fee }} / {{ v3_fee_precision }} as token0_fee_rate,
+            token1,
+            {{ v3_default_fee }} / {{ v3_fee_precision }} as token1_fee_rate
+        from {{ source("camelot_v3_arbitrum", "AlgebraFactory_evt_Pool") }}
+    ),
+    v3_directional_fee_rate_updates as (
+        select
+            date_trunc('minute', evt_block_time) as minute,
+            pair,
+            version,
+            token0,
+            feezto / {{ v3_fee_precision }} as token0_fee_rate,
+            token1,
+            feeotz / {{ v3_fee_precision }} as token1_fee_rate
+        from {{ source("camelot_v3_arbitrum", "AlgebraPool_evt_Fee") }} as fee_updates
+        join
+            v3_pairs_with_initial_fee_rates as pairs
+            on fee_updates.contract_address = pairs.pair
+    ),
+    pairs as (
+        select *
+        from v2_pairs_with_initial_fee_rates
+        union all
+        select *
+        from v3_pairs_with_initial_fee_rates
+    ),
+    fee_rate_updates as (
+        select
+            minute,
+            pair,
+            version,
+            token0,
+            avg(token0_fee_rate) as token0_fee_rate,  -- Handle edge case where pair fees gets changed multiple times per minute
+            token1,
+            avg(token1_fee_rate) as token1_fee_rate  -- Handle edge case where pair fees gets changed multiple times per minute
+        from
+            (
+                select *
+                from pairs
+                union all
+                select *
+                from v2_directional_fee_rate_updates
+                union all
+                select *
+                from v3_directional_fee_rate_updates
+            )
+        group by minute, pair, version, token0, token1
+    ),
+    camelot_pair_trades_by_minute as (
+        select distinct
+            date_trunc('minute', block_time) as minute, project_contract_address as pair
+        from dex.trades
+        where
+            blockchain = '{{blockchain}}' and project = 'camelot'
+            {% if not is_incremental() %}
+                and block_time >= timestamp '{{project_start_date}}'
+            {% endif %}
+            {% if is_incremental() %}
+                and block_time >= date_trunc('day', now() - interval '7' day)
+            {% endif %}
+    ),
+    -- Prepare data structure (1 row for every minute where pair trades happened
+    -- and/or fee rates got updated)
+    pairs_by_minute as (
+        select distinct minute, pair
+        from
+            (
+                select minute, pair
+                from camelot_pair_trades_by_minute
+                union all
+                select minute, pair
+                from fee_rate_updates
+            )
+    ),
```

Comment:
```suggestion
    pairs_by_minute as (
          select minute, pair
          from camelot_pair_trades_by_minute
          union distinct
          select minute, pair
          from fee_rate_updates
    ),
```

Reply:
i think where i got confused was this union all then select distinct -- i wonder if this simplified approach works?

Reply:
i didn't know about union distinct, but from naming alone it sounds like it would produce the desired result!
can we add some kind of tests to this spellbook also? so we can see if these changes would change the resultset?

Reply:
what you can do:
- run CI in current setup, have CI tables ready to query on the dune app
- push a commit to PR, rerun CI on commit, query new CI tables (should have unique git sha value per commit, therefore unique tables)

you can then compare the results with the `exclude` function or just manually